sdk/trace/span.go

```diff
@@ -426,6 +427,15 @@ func (s *recordingSpan) End(options ...trace.SpanEndOption) {
 	s.mu.Unlock()
 
 	sps := s.tracer.provider.getSpanProcessors()
+	for _, sp := range sps {
+		if oesp, ok := sp.sp.(x.OnEndingSpanProcessor); ok {
+			oesp.OnEnding(s)
+		}
+	}
```

Comment:
Can't this be run before setting the end time?

That would avoid needing the extra `hasEnded` field and duplicate lock acquiring.

Reply:
No, as the spec says:

> `OnEnding` is called during the span `End()` operation, after the end timestamp has been set.